osutil/syncdir.go

```diff
@@ -143,42 +187,101 @@ func matchAny(globs []string, path string) (ok bool, index int, err error) {
 // EnsureDirState ensures that directory content matches expectations.
 //
 // This is like EnsureDirStateGlobs but it only supports one glob at a time.
-func EnsureDirState(dir string, glob string, content map[string]*FileState) (changed, removed []string, err error) {
+func EnsureDirState(dir string, glob string, content map[string]FileState) (changed, removed []string, err error) {
 	return EnsureDirStateGlobs(dir, []string{glob}, content)
 }
 
-// Equals returns whether the file exists in the expected state.
-func (fileState *FileState) Equals(filePath string) (bool, error) {
-	stat, err := os.Stat(filePath)
+// fileStateEqualTo returns whether the file exists in the expected state.
+func fileStateEqualTo(filePath string, state FileState) (bool, error) {
+	other := &FileReference{Path: filePath}
+
+	// Open views to both files so that we can compare them.
+	readerA, sizeA, modeA, err := state.State()
+	if err != nil {
+		return false, err
+	}
+	defer readerA.Close()
+
+	readerB, sizeB, modeB, err := other.State()
 	if err != nil {
 		if os.IsNotExist(err) {
-			// not existing is not an error
+			// Not existing is not an error
 			return false, nil
 		}
 		return false, err
 	}
-	if stat.Mode().Perm() == fileState.Mode.Perm() && stat.Size() == int64(len(fileState.Content)) {
-		content, err := ioutil.ReadFile(filePath)
-		if err != nil {
-			return false, err
+	defer readerB.Close()
+
+	// If the files have different size or different mode they are not
+	// identical and need to be re-created. Mode change could be optimized to
+	// avoid re-writing the whole file.
+	if modeA.Perm() != modeB.Perm() || sizeA != sizeB {
+		return false, nil
+	}
+	// The files have the same size so they might be identical.
+	// Do a block-wise comparison to determine that.
+	return StreamEqual(readerA, readerB, 0)
+}
+
+// StreamEqual returns true if both streams have same length and content.
+func StreamEqual(readerA, readerB io.Reader, chunkSize int) (bool, error) {
+	if readerA == readerB {
+		return true, nil
+	}
+	if chunkSize <= 0 {
+		chunkSize = 4096
+	}
+	bufA := make([]byte, chunkSize)
+	bufB := make([]byte, chunkSize)
+	for {
+		nA, errA := readerA.Read(bufA)
+		toReadFromB := nA
+		if toReadFromB == 0 {
+			// If we read nothing from stream A we want to get a chance to read
+			// something from B so that we can detect streams of unequal
+			// length.
+			toReadFromB = 1
+		}
+		nB, errB := io.ReadAtLeast(readerB, bufB, toReadFromB)
+		// We read the same non-empty amount from each stream.
+		if nA == nB && nA > 0 {
+			if bytes.Equal(bufA[:nA], bufB[:nB]) {
+				continue
+			}
+		}
+		// We read nothing from both streams.
+		if nA == nB && nA == 0 {
```

Comment:
I think you should consider the errors before handling `nA == 0 && nB == 0`, otherwise when both readers fail and return 0, <err>, the code will return true

Reply:
This code is gone now.